### DISCUSSION TITLE
[과목 2] 제 1장 SQL 기본

### DISCUSSION THREAD

**Discussion by beolggoolosori** [flataex/sqlp-study#13] [Custom 연습문제] [answered]
1. 다음 중 SELECT 문에서 WHERE 절과 HAVING 절의 주요 차이점은 무엇인가요?
① WHERE 절은 행을 필터링하는 데 사용되고, HAVING 절은 열을 필터링하는 데 사용된다.
② HAVING 절은 행을 필터링하는 데 사용되고, WHERE 절은 열을 필터링하는 데 사용된다.
③ WHERE 절은 GROUP BY 절 이전에 행을 필터링하는 데 사용되고, HAVING 절은 GROUP BY 절 이후에 집계된 결과를 필터링하는 데 사용된다.
④ HAVING 절은 GROUP BY 절 이전에 행을 필터링하는 데 사용되고, WHERE 절은 GROUP BY 절 이후에 집계된 결과를 필터링하는 데 사용된다.

2. 다음 중 SQL에서 JOIN 절의 역할은 무엇인가요?
① 두 개 이상의 테이블을 기반으로 새로운 테이블을 생성한다.
② 한 테이블의 행을 필터링하는 데 사용된다.
③ 두 개 이상의 테이블을 기반으로 한 테이블에서 관련 열을 결합하여 쿼리 결과를 생성한다.
④ 두 개 이상의 테이블에서 공통 열을 제거한다.

3. happysql database의 `ADMIN.EMPLOYEE` 테이블과 `ADMIN.DEPARTMENT`테이블을 참고하여 아래 제시된 문제에 대한 SQL 및 쿼리결과를 작성시오.

>각 부서별로 급여 평균이 50,000보다 높은 직원들을 가진 부서를 찾으세요. 결과는 부서명과 해당 부서의 급여 평균을 포함해야 합니다.

>평균 급여를 내림차순으로 정렬하고, 평균 급여가 높은 상위 3개 부서의 정보만 출력하세요.

**Comment by white-gyu** [comment, marked as answer]
1. 다음 중 SELECT 문에서 WHERE 절과 HAVING 절의 주요 차이점은 무엇인가요?
① WHERE 절은 행을 필터링하는 데 사용되고, HAVING 절은 열을 필터링하는 데 사용된다.
② HAVING 절은 행을 필터링하는 데 사용되고, WHERE 절은 열을 필터링하는 데 사용된다.
③ WHERE 절은 GROUP BY 절 이전에 행을 필터링하는 데 사용되고, HAVING 절은 GROUP BY 절 이후에 집계된 결과를 필터링하는 데 사용된다.
④ HAVING 절은 GROUP BY 절 이전에 행을 필터링하는 데 사용되고, WHERE 절은 GROUP BY 절 이후에 집계된 결과를 필터링하는 데 사용된다.

> ③

2. 다음 중 SQL에서 JOIN 절의 역할은 무엇인가요?
① 두 개 이상의 테이블을 기반으로 새로운 테이블을 생성한다.
② 한 테이블의 행을 필터링하는 데 사용된다.
③ 두 개 이상의 테이블을 기반으로 한 테이블에서 관련 열을 결합하여 쿼리 결과를 생성한다.
④ 두 개 이상의 테이블에서 공통 열을 제거한다.

> ③

3. happysql database의 `ADMIN.EMPLOYEE` 테이블과 `ADMIN.DEPARTMENT`테이블을 참고하여 아래 제시된 문제에 대한 SQL 및 쿼리결과를 작성시오.

>각 부서별로 급여 평균이 50,000보다 높은 직원들을 가진 부서를 찾으세요. 결과는 부서명과 해당 부서의 급여 평균을 포함해야 합니다.

>평균 급여를 내림차순으로 정렬하고, 평균 급여가 높은 상위 3개 부서의 정보만 출력하세요.

```sql
select *
from (select DEPARTMENT.id, DEPARTMENT_NAME, avg(SALARY)
      from DEPARTMENT
               join EMPLOYEE E on DEPARTMENT.ID = E.DEPARTMENT_ID
      group by DEPARTMENT.ID, DEPARTMENT_NAME
      having avg(SALARY) > 50000
      order by avg(SALARY) desc)
where ROWNUM <= 3;
```

**Comment by Psjepark** [comment]
1. 3
2. 3
```SQL
use admin;
go;

select  top 3 A.department_name,  avg(B.salary)
from department A 
inner join employee B
on A.id = B.department_id
where  avg(B.salary) > 50000
group by A.department_name, B.salary
order by  avg(B.salary);
```

**Comment by white-gyu** [comment]
[oracle top 함수 구현하기](https://funfunit.tistory.com/192)

**Comment by beolggoolosori** [comment]
```sql
SELECT department_name,
       avg_salary
FROM (SELECT d.department_name,
             AVG(e.salary)                                   AS avg_salary,
             ROW_NUMBER() OVER (ORDER BY AVG(e.salary) DESC) AS row_num
      FROM employee e
               JOIN department d ON e.department_id = d.id
      GROUP BY d.department_name
      HAVING AVG(e.salary) > 50000)
WHERE row_num <= 3;
